### PR TITLE
18-10-01_make-train-val-dataset

### PR DESCRIPTION
**Pull request by estepona**
Hi! This PR adds one function `make_train_val_dataset` to `torchvision/datasets/folder.py`. I'm creating this PR because I've been having this problem for a while, and it finally occurs to me that maybe I should contribute my solution to other people who might have faced the same problem like myself.

### Problem
When trying to load my own training and validation datasets organized in folders, the native combination of `DatasetFolder`, `ImageFolder`, and `DataLoader` is great, as long as I put pictures arranged as the following before loading:
>train/dog/xxx.png
>train/dog/xxy.png
>train/cat/123.png
>train/cat/nsdf3.png
>val/cat/asd932_.png
>val/dog/xxz.png

While most of the datasets downloaded are arranged in this way:
>root/dog/xxx.png
>root/dog/xxy.png
>root/dog/xxz.png
>root/cat/123.png
>root/cat/nsdf3.png
>root/cat/asd932_.png

I have to do the file structure thing beforehand because `DatasetFolder` and `ImageFolder` reads all classes and files under the `root` folder, creating a `train` and a `val` `DatasetFolder` instance means setting up those two as separate root folders and copying/moving/splitting subfolders and files under them. Having a tiny dataset, I can do it manually. Having a dataset of more than a few hundreds pictures then becomes annoying. I have to write custom functions to copy folder and files into a location that are arranged like `<train>or<val>/<class>/`. Having a dataset with dozens of GB or even TB would take a lot of I/O time, as well as extra disk space. Recently I am working on a [Places205](http://places.csail.mit.edu/downloadCNN.html) dataset that contains 2.5 million images. Downloading the 256x256 dataset would take 132GB of disk space. For prototyping, copying only 10 classes (out of 205 classes) for training takes nearly 2 minutes before training starts, which sometimes interrupts the overall flow. 

### Solution
So I decided to read the source code in `folder.py`, and found that I don't need to copy everything in a new location, instead, I only need to modify three attributes of `DatasetFolder`: `classes`, `class_to_idx`, and `samples`, especially `samples`, which has the path and class index of each file. So I added this function `make_train_val_dataset` to alter the three attributes, with additional arguments to specify the classes to train and the train ratio. I don't want to modify other classes or functions in `folder.py` because I think they are well-designed. I only introduced this function and imported `random` for shuffling. 

### Usage
For example, if you have a dataset that looks like this, with 4 classes and 10 images for each:
>root/dog/1.png
>root/dog/2.png
>root/dog/3.png
>root/dog/4.png
>root/dog/5.png
>root/dog/6.png
>root/dog/7.png
>root/dog/8.png
>root/dog/9.png
>root/dog/10.png
>root/cat/1.png
>......
>root/tiger/1.png
>......
>root/lion/1.png
>......

If you only want to train a classifier to classify dog and cat, with 80% pictures for training (default is 95%), and 20% for validation, you only need to call something like:  
`train_ds, val_ds = make_train_val_dataset(root=os.path.join('resources', 'img'), classes=['dog', 'cat'], train_ratio=0.8)`  
Then if you check `train_ds`, you get:  
>Dataset DatasetFolder
>&nbsp;&nbsp;Number of datapoints: 16
>&nbsp;&nbsp;Root Location: resources\img
>&nbsp;&nbsp;Transforms (if any): None
>&nbsp;&nbsp;Target Transforms (if any): None
>['cat', 'dog']
>{'cat': 0, 'dog': 1}
>[('resources\\img\\cat\\4.jpg', 0), ('resources\\img\\cat\\1.jpg', 0), ('resources\\img\\cat\\10.jpg', 0), ('resources\\img\\cat\\8.jpg', 0), ('resources\\img\\cat\\3.jpg', 0), ('resources\\img\\cat\\5.jpg', 0), ('resources\\img\\cat\\2.jpg', 0), ('resources\\img\\cat\\6.jpg', 0), ('resources\\img\\dog\\10.jpg', 1), ('resources\\img\\dog\\8.jpg', 1), ('resources\\img\\dog\\6.jpg', 1), ('resources\\img\\dog\\7.jpg', 1), ('resources\\img\\dog\\4.jpg', 1), ('resources\\img\\dog\\1.jpg', 1), ('resources\\img\\dog\\9.jpg', 1), ('resources\\img\\dog\\2.jpg', 1)]

and for `val_ds`:  
>Dataset DatasetFolder
>&nbsp;&nbsp;Number of datapoints: 4
>&nbsp;&nbsp;Root Location: resources\img
>&nbsp;&nbsp;Transforms (if any): None
>&nbsp;&nbsp;Target Transforms (if any): None
>['cat', 'dog']
>{'cat': 0, 'dog': 1}
>[('resources\\img\\cat\\7.jpg', 0), ('resources\\img\\cat\\9.jpg', 0), ('resources\\img\\dog\\3.jpg', 1), ('resources\\img\\dog\\5.jpg', 1)]

Then you can safely use `DataLoader` to create batches and(or) transform.

I hope this is helpful. If you have any feedback, I'd love to hear. 

This is my first PR, I'm excited. Thank you for developing PyTorch and Torchvision. They are awesome libraries.

### CODE REVIEW

**Comment by estepona** [comment]
I found an unexpected error when class indexes are not from 0 to n-1. I've closed this PR.
